Reconstruction/RecCalorimeter/src/components/SingleCaloTowerTool.cpp

```diff
@@ -41,96 +43,134 @@ StatusCode SingleCaloTowerTool::initialize() {
     error() << "There is no phi-eta segmentation." << endmsg;
     return StatusCode::FAILURE;
   }
-  if (!m_etaMax) {
-    warning() << "Undefined detector size in eta. In each event the cell collection will be searched for maximum eta."
-              << endmsg;
-  }
   return StatusCode::SUCCESS;
 }
 
 StatusCode SingleCaloTowerTool::finalize() { return GaudiTool::finalize(); }
 
 tower SingleCaloTowerTool::towersNumber() {
+  // maximum eta of the detector (== eta offset + half of the cell size)
+  m_etaMax = fabs(m_segmentation->offsetEta() - m_segmentation->gridSizeEta() * 0.5);
+  m_phiMax = fabs(m_segmentation->offsetPhi() - M_PI / (double)m_segmentation->phiBins());
+
   // number of phi bins
-  m_nPhiTower = idPhi(2 * M_PI);
+  float epsilon = 0.0001;
+  m_nPhiTower = ceil(2 * (m_phiMax - epsilon) / m_deltaPhiTower);
   // number of eta bins (if eta maximum is defined)
-  if (m_etaMax) {
-    m_nEtaTower = 2 * idEta(m_etaMax - m_deltaEtaTower / 2.) + 1;
-  } else {
-    m_nEtaTower = 0;
-  }
+  m_nEtaTower = ceil(2 * (m_etaMax - epsilon) / m_deltaEtaTower);
+  debug() << "etaMax " << m_etaMax << ", deltaEtaTower " << m_deltaEtaTower << ", nEtaTower " << m_nEtaTower << endmsg;
+  debug() << "phiMax " << m_phiMax << ", deltaPhiTower " << m_deltaPhiTower << ", nPhiTower " << m_nPhiTower << endmsg;
+
   tower total;
   total.eta = m_nEtaTower;
   total.phi = m_nPhiTower;
   return total;
 }
 
-int SingleCaloTowerTool::etaTowersNumber() {
-  // recalculate number of eta bins for each event
-  const fcc::CaloHitCollection* cells = m_cells.get();
-  float etaCell = 0;
-  for (const auto& cell : *cells) {
-    etaCell = fabs(m_segmentation->eta(cell.core().cellId));
-    if (etaCell > m_etaMax) {
-      m_etaMax = etaCell;
-    }
-  }
-  // eta from cell collection is middle of cell
-  m_nEtaTower = 2 * m_etaMax / m_deltaEtaTower + 1;
-  return m_nEtaTower;
-}
-
 uint SingleCaloTowerTool::buildTowers(std::vector<std::vector<float>>& aTowers) {
-  // Get the input collection with cells from simulation + digitisation (after calibration and with noise)
+  // Get the input collection with cells from simulation + digitisation (after
+  // calibration and with noise)
   const fcc::CaloHitCollection* cells = m_cells.get();
   debug() << "Input Hit collection size: " << cells->size() << endmsg;
   // Loop over a collection of calorimeter cells and build calo towers
-  int iPhi = 0, iEta = 0;
-  float etaCell = 0;
+  // borders of the cell in eta/phi
+  float etaCellMin, etaCellMax;
+  float phiCellMin, phiCellMax;
+  // tower index of the borders of the cell
+  int iPhiMin = 0, iPhiMax = 0;
+  int iEtaMin = 0, iEtaMax = 0;
+  // fraction of cell in eta/phi belonging a tower
```

Comment:
I would add an extended comment what happens if cell is smaller than tower (all fractions are equal to 1 because first==middle==last tower ?) and what if it is larger (sum of fractions == 1). Maybe also that fractions are calculated based on the size?